src/custom-element.js

```diff
@@ -123,92 +123,103 @@ export function stubElements(win) {
  * @param {!AmpElement} element
  */
 export function applyLayout_(element) {
-  let widthAttr = element.getAttribute('width');
-  let heightAttr = element.getAttribute('height');
-  const sizesAttr = element.getAttribute('sizes');
   const layoutAttr = element.getAttribute('layout');
+  const widthAttr = element.getAttribute('width');
+  const heightAttr = element.getAttribute('height');
+  const sizesAttr = element.getAttribute('sizes');
+
+  // Input layout attributes.
+  const inputLayout = layoutAttr ? parseLayout(layoutAttr) : null;
+  assert(inputLayout || !layoutAttr, 'Unknown layout: %s', layoutAttr);
+  const inputWidth = widthAttr && widthAttr != 'auto' ?
+      parseLength(widthAttr) : widthAttr;
+  assert(inputWidth || !widthAttr, 'Invalid width value: %s', widthAttr);
+  const inputHeight = heightAttr ? parseLength(heightAttr) : null;
+  assert(inputHeight || !heightAttr, 'Invalid height value: %s', heightAttr);
+
+  // Effective layout attributes. These are effectively constants.
+  let width;
+  let height;
+  let layout;
 
-  // Handle elements that do not specify a width/height and are defined to have
-  // natural browser dimensions.
-  if ((!layoutAttr || layoutAttr == Layout.FIXED ||
-          layoutAttr == Layout.FIXED_HEIGHT) &&
-      (!widthAttr || !heightAttr) && hasNaturalDimensions(element.tagName)) {
+  // Calculate effective width and height.
+  if ((!inputLayout || inputLayout == Layout.FIXED ||
+          inputLayout == Layout.FIXED_HEIGHT) &&
+      (!inputWidth || !inputHeight) && hasNaturalDimensions(element.tagName)) {
+    // Default width and height: handle elements that do not specify a
+    // width/height and are defined to have natural browser dimensions.
     const dimensions = getNaturalDimensions(element.tagName);
-    if (layoutAttr != Layout.FIXED_HEIGHT) {
-      widthAttr = widthAttr || dimensions.width;
-    }
-    heightAttr = heightAttr || dimensions.height;
+    width = inputWidth || inputLayout == Layout.FIXED_HEIGHT ? inputWidth :
+        dimensions.width;
+    height = inputHeight || dimensions.height;
+  } else {
+    width = inputWidth;
+    height = inputHeight;
   }
 
-  let layout;
-  if (layoutAttr) {
-    // TODO(dvoytenko): show error state visually in the dev mode, e.g.
-    // red background + error message.
-    layout = parseLayout(layoutAttr.trim());
-    if (!layout) {
-      throw new Error('Unknown layout: ' + layoutAttr);
-    }
-  } else if (widthAttr || heightAttr) {
-    if (!widthAttr || widthAttr == 'auto') {
-      layout = Layout.FIXED_HEIGHT;
+  // Calculate effective layout.
+  if (inputLayout) {
+    // Layout specified directly.
+    layout = inputLayout;
+  } else {
```

Comment:
Here is a way to fold this out so for each layout it says maybe a tad more directly what the condition is (the order still matters, so not claiming it's completely pure).

else if (!width && !height) {
  layout = Layout.CONTAINER;
} else if (height && (!width || width=='auto')) {
  layout = Layout.FIXED_HEIGHT;
} else if (height && width && !sizesAttr) {
  layout = Layout.RESPONSIVE;
} else {
  layout = Layout.FIXED;
}